Improve DisplayASMCode and add DisplayCCode

mk-soft-65 opened the pull request:
- Add missing editor constants
- Show error message when started outside of the PB IDE
- Add support for PB 6.0
- Keep ASM file existing and display path in the output
- Transfer no tool header to standard program / clipboard
- Add DisplayCCode

mk-soft-65 commented:
Sorry,

bad filename (Case)

SicroAtGit commented:
Thank you for the PR.

I saw [your post](https://www.purebasic.fr/german/viewtopic.php?p=361093#p361093) in the German PureBasic forum, but didn't have enough time to look at it in detail until now.

> * Check and correction of environment variable "PB_TOOL_Compiler"

I wonder if it doesn't make more sense if we wait until this [PB_TOOL_Compiler update delayed](https://www.purebasic.fr/english/viewtopic.php?t=72110) bug in the PureBasic IDE is fixed. Then we could rename this tool to _Display-ASM/C-Code_ or something like that and keep everything in one tool.

If I have enough time, I will have a look at the corresponding PureBasic IDE code and try to fix the bug.

> * PB Editor standard constants added

Thanks.

> * Output to standard editor and clipboard the asmCode without own header

We could add a checkbox gadget for this to make it optional (default is then all content, i.e. checkbox gadget is set). I think that some users want to take the whole content that is displayed in the textbox gadget.

> * Include files dependencies resolved

I prefer to keep dependencies external to avoid duplicate code and to only have to make improvements to the dependency code once in one place for each dependency.

SicroAtGit commented:
I have now created a [pull request with the `PB_TOOL_Compiler` fix](https://github.com/fantaisie-software/purebasic/pull/191) in the PureBasic IDE repository.

SicroAtGit commented on PB-IDE-Tools/displayASMcode.pb:
```diff
@@ -0,0 +1,308 @@
+; 
+; Author        : Sicro
+; Author second : mk-soft
+; Update        : 28.11.2021
```
I don't like the author, author second and update lines and the minus character before word _Description_.

SicroAtGit commented on PB-IDE-Tools/displayASMcode.pb:
```diff
@@ -0,0 +1,308 @@
+
+; MIT License
+;
+; Copyright (c) 2018-2020 Sicro
```
```suggestion
; Copyright (c) 2018-2020 Sicro
; Copyright (c) 2022 mk-soft
```

SicroAtGit commented on PB-IDE-Tools/displayASMcode.pb:
```diff
@@ -0,0 +1,308 @@
+;-Include Codes
+; =============
+
+Procedure.i OpenWithStandardProgram(FilePath$)
```
Dependencies should not be inserted directly.

SicroAtGit commented on PB-IDE-Tools/displayASMcode.pb:
```diff
@@ -0,0 +1,308 @@
+  
+EndProcedure
+
+;XIncludeFile "../File/GetFileContentAsString.pbi"
```
Dependencies should not be inserted directly.

SicroAtGit commented on PB-IDE-Tools/displayASMcode.pb:
```diff
@@ -0,0 +1,308 @@
+
+; Read the ASM code file
+file = ReadFile(#PB_Any, asmCodeFilePath$)
+If file
```
Dependencies should not be inserted directly.

SicroAtGit commented on PB-IDE-Tools/displayCcode.pb:
```diff
@@ -0,0 +1,310 @@
+; Author        : Sicro
+; Author second : mk-soft (Modified for C-Code)
+; Update        : 28.11.2021
+
```
I don't like the author, author second and update lines and the minus character before word _Description_.

SicroAtGit commented on PB-IDE-Tools/displayCcode.pb:
```diff
@@ -0,0 +1,310 @@
+
+; MIT License
+;
+; Copyright (c) 2018-2020 Sicro
```
```suggestion
; Copyright (c) 2018-2020 Sicro
; Copyright (c) 2022 mk-soft
```

SicroAtGit commented on PB-IDE-Tools/displayCcode.pb:
```diff
@@ -0,0 +1,310 @@
+;-Include Codes
+; =============
+
+Procedure.i OpenWithStandardProgram(FilePath$)
```
Dependencies should not be inserted directly.

SicroAtGit commented on PB-IDE-Tools/displayCcode.pb:
```diff
@@ -0,0 +1,310 @@
+  
+EndProcedure
+
+;XIncludeFile "../File/GetFileContentAsString.pbi"
```
Dependencies should not be inserted directly.

SicroAtGit commented on PB-IDE-Tools/displayCcode.pb:
```diff
@@ -0,0 +1,310 @@
+
+; Read the ASM code file
+file = ReadFile(#PB_Any, asmCodeFilePath$)
+If file
```
Dependencies should not be inserted directly.

SicroAtGit reviewed (changes requested):
Rename the code files:

- `displayASMcode.pb` => `DisplayASMCode.pb`
- `displayCcode.pb` => `DisplayCCode.pb`

SicroAtGit commented:
@mk-soft-65 I am still unsure if it should be two tools. Is it useful for you to have two separate tools?

SicroAtGit commented on PB-IDE-Tools/DisplayASMCode.pb:
```diff
@@ -1,8 +1,13 @@
+; Author        : Sicro
+; Author second : mk-soft
+; Update        : 28.11.2021
+
```
I don't like this. Add a copyright line for you in the license instead:
```
; MIT License
;
; Copyright (c) 2018-2020 Sicro
; Copyright (c) 2021-2022 mk-soft
```

SicroAtGit commented on PB-IDE-Tools/DisplayASMCode.pb:
```diff
@@ -39,8 +44,29 @@ EnableExplicit
+  
+EndProcedure
+
+;XIncludeFile "../File/GetFileContentAsString.pbi"
```
Include file content should not be inserted. Please reset this change to the original code:
```
XIncludeFile "../System/OpenStandardProgram.pbi"
XIncludeFile "../File/GetFileContentAsString.pbi"
```

SicroAtGit commented on PB-IDE-Tools/DisplayASMCode.pb:
```diff
@@ -166,7 +230,12 @@ Else
 EndIf
 
-asmCode$ = GetFileContentAsString(asmCodeFilePath$)
+; Read the ASM code file
+file = ReadFile(#PB_Any, asmCodeFilePath$)
+If file
+  asmCode$ + ReadString(file, #PB_File_IgnoreEOL)
+  CloseFile(file)
+EndIf
```
Please reset this code to the original code:
```
asmCode$ = GetFileContentAsString(asmCodeFilePath$)
```

SicroAtGit commented on PB-IDE-Tools/displayCcode.pb:
```diff
@@ -0,0 +1,340 @@
+; English-Forum :
+; French-Forum  :
+; German-Forum  : https://www.purebasic.fr/german/viewtopic.php?f=10&t=30935
+; 
+; Author        : Sicro
+; Author second : mk-soft (Modified for C-Code)
+; Update        : 28.11.2021
+
```
I don't like this. Add a copyright line for you in the license instead:
```
; MIT License
;
; Copyright (c) 2018-2020 Sicro
; Copyright (c) 2021-2022 mk-soft
```

SicroAtGit commented on PB-IDE-Tools/displayCcode.pb:
```diff
@@ -0,0 +1,340 @@
+;-Include Codes
+; =============
+
+Procedure.i OpenWithStandardProgram(FilePath$)
+  
+  Protected Result
+  
+  ; Avoid problems with paths containing spaces
+  FilePath$ = #DQUOTE$ + FilePath$ + #DQUOTE$
+  
+  CompilerSelect #PB_Compiler_OS
+    CompilerCase #PB_OS_Windows
+      ; https://docs.microsoft.com/en-us/windows/desktop/api/shellapi/nf-shellapi-shellexecutew
+      Result = Bool(ShellExecute_(0, "open", FilePath$, #Null, #Null, #SW_SHOW) > 32)
+    CompilerCase #PB_OS_Linux
+      ; https://portland.freedesktop.org/doc/xdg-open.html
+      Result = Bool(RunProgram("xdg-open", FilePath$, GetCurrentDirectory()))
+    CompilerCase #PB_OS_MacOS
+      Result = Bool(RunProgram("open", FilePath$, GetCurrentDirectory()))
+  CompilerEndSelect
+  
+  ProcedureReturn Result
+  
+EndProcedure
+
+;XIncludeFile "../File/GetFileContentAsString.pbi"
```
Include file content should not be inserted. Please reset this change to the original code:
```
XIncludeFile "../System/OpenStandardProgram.pbi"
XIncludeFile "../File/GetFileContentAsString.pbi"
```

SicroAtGit commented on PB-IDE-Tools/displayCcode.pb:
```diff
@@ -0,0 +1,340 @@
+;-Define Local Variables
+; ======================
+
+Define compilerHomePath$, compilerFilePath$, workingDirectoryPath$, codeFilePath$, codeTempFilePath$,
+       asmCodeFilePath$, exeFilePath$, outputFilePathForStandardProgram$, compilerParameters$,
+       compilerUserParameters$
+Define asmCode$, output$, compilerOutput$
```
Please change:
- `asmCodeFilePath$` to `cCodeFilePath$`
- `asmCode$` to `cCode$`

SicroAtGit commented on PB-IDE-Tools/displayCcode.pb:
```diff
@@ -0,0 +1,340 @@
+DeleteFile(exeFilePath$)
+
+; ====================
+;-Create ASM Code File
```
Please replace `ASM` with `C`. There are more places in which this should be done, too.

SicroAtGit commented on PB-IDE-Tools/displayCcode.pb:
```diff
@@ -0,0 +1,340 @@
+  isCompilerError = #True
+EndIf
+
+; Read the ASM code file
+file = ReadFile(#PB_Any, asmCodeFilePath$)
+If file
+  asmCode$ + ReadString(file, #PB_File_IgnoreEOL)
+  CloseFile(file)
+EndIf
```
Please reset this code to the original code:
```
cCode$ = GetFileContentAsString(cCodeFilePath$)
```

mk-soft-65 committed:
DisplayASMCode: Add missing editor constants

mk-soft-65 committed:
DisplayASMCode: Show error message when started outside of the PB IDE

mk-soft-65 committed:
DisplayASMCode: Add support for PB 6.0

The PB compiler parameter for constants has changed.

If a PB compiler C backend was specified, it now switches to the PB
compiler ASM backend.

mk-soft-65 committed:
DisplayASMCode: Keep ASM file existing and display path in the output

mk-soft-65 committed:
DisplayASMCode: Transfer no tool header to standard program / clipboard

mk-soft-65 committed:
Add DisplayCCode

SicroAtGit commented:
I have now made the remaining changes and merged the commits into meaningful commits.

@mk-soft-65 Thanks a lot for your PR!
